### PR TITLE
[WIP] Documentation efforts

### PR DESCRIPTION
**Pull request by adamczykm**

### CODE REVIEW

File: src/utils/fp/TaskEither.ts [hunk 1]
```diff
@@ -49,12 +64,19 @@ export function dropResult<E, T>(t: TaskEither<E, T>): TaskEither<E, void> {
   return TE.map(() => undefined)(t);
 }
```
Comment (chfanghr) [inline src/utils/fp/TaskEither.ts:65]: Let's drop this in favor of [`asUnit`](https://gcanti.github.io/fp-ts/modules/TaskEither.ts.html#asunit).

File: src/utils/fp/TaskEither.ts [hunk 2]
```diff
@@ -80,20 +105,31 @@ export function safeFromString<T>(
...
   key: string,
   params: { [key: string]: string }
 ): TaskEither<string, Field> {
   return getParam(safeFromString(Field))(key, params);
 }
 
+/**
+ * Converts a promise that may resolve into an empty value into
+ * a TaskEither object that fails with a message if the promise resolves into an empty value
+ * or is rejected, or throws an exception.
```
Comment (chfanghr) [inline src/utils/fp/TaskEither.ts:131]:
```suggestion
 * a TaskEither object that fails with a message if the promise resolves into an empty value,
 * is rejected, or throws an exception.
```

File: src/library/tools/pluginServer/index.ts [hunk 3]
```diff
@@ -79,4 +94,15 @@ const main = pipe(
   })
 );
 
+/**
+ * The input to /verifyProof
+ *  TODO switch to zod schema.
```
Comment (chfanghr) [inline src/library/tools/pluginServer/index.ts:104]: Yeah good idea, not sure how we're gonna deal with `JsonProof` though.

File: src/library/server/minauthStrategy.ts [hunk 4]
```diff
@@ -1,3 +1,12 @@
+// TODO this module is very provisional
```
Comment (chfanghr) [inline src/library/server/minauthStrategy.ts:1]: It's been revised in my pr: https://github.com/mlabs-haskell/MinAuth/blob/846245499c8c951920b5dc27b134633d60c5f4d6/src/library/server/minauthStrategy.ts.

The new version should meet most of the requirements listed below, modulo the composability.

File: src/library/plugin/index.ts [hunk 5]
```diff
@@ -50,3 +63,7 @@ export const verifyProof = (
       pluginName
     )
   );
+
+// TODO DEV NOTE:
+// the lib/plugin modules should be required to build a plugin.
+// The stuff that is needed to operate them should be exported from the server library.
```
Comment (chfanghr) [inline src/library/plugin/index.ts:72]: You mean keeping them in a separate package?

File: src/library/tools/pluginServer/index.ts [hunk 6]
```diff
@@ -11,34 +11,48 @@ import { Either } from 'fp-ts/Either';
...
-  proof: JsonProof;
-}
-
+/**
+ * This function starts a provisional plugin server meant mostly for development purposes,
+ * so that developing your backend services doesn't require restarting plugins and their services.
+ * It reads configuration from a file and initializes a set of plugins.
+ * You can interact with the plugins via HTTP API consisting of:
+ * - '/verifyProof' endpoint to validate proofs and get a plugin's output
+ * - TODO '/checkOutputValidity' endpoint to verify if output is still valid
```
Comment (chfanghr) [inline src/library/tools/pluginServer/index.ts:21]: Good catch, I think it was lost in rebasing.

File: src/library/plugin/index.ts [hunk 7]
```diff
@@ -9,36 +9,49 @@ export * from './fp/pluginType';
...
+ */
 export const installCustomRoutes = (
   activePlugins: ActivePlugins,
   app: expressCore.Express
 ): Promise<void> => launchTE(fpUtils.installCustomRoutes(activePlugins)(app));
 
+/**
+ * Given set of plugins, plugin identifier and plugin proof * verification inputs
+ * - verify proof, and store and return the plugin output.
+ */
```
Comment (chfanghr) [inline src/library/plugin/index.ts:51]: The format was messed up in copy-paste

File: src/plugins/merkleMemberships/server/treeStorage.ts
Comment (chfanghr) [inline src/plugins/merkleMemberships/server/treeStorage.ts:1]: Should have called `TreeProvider` `Forest`. Naming is hard.

File: src/library/plugin/fp/pluginLoader.ts [hunk 8]
```diff
@@ -27,56 +27,185 @@ import {
...
   fpInterfaceTag,
   tsInterfaceTag
-} from './interfaceKind';
+} from '@lib/common/interfaceKind';
 
+/**
+ * Configuration schema for the plugin loader
+ */
 export const configurationSchema = z.object({
+  // Directory where to look for plugins
```
Comment (chfanghr) [inline src/library/plugin/fp/pluginLoader.ts:36]: More like a fallback location - If the `path` of a plugin is specified `pluginDir` won't be taken into consideration.

File: src/library/plugin/fp/proofCache.ts [hunk 9]
```diff
@@ -15,62 +15,112 @@ import {
...
+/**
+ * A key that cache uses to efficiently identify stored proofs.
+ */
 export type ProofKey = string;
 
+// TODO: is my understanding correct that:
+// checkEachProof: CheckCachedProofs<I>;
+// is meant to validate the entire cache given a validator for proofs?
+// And all proofs that are left after this operation should be regarded as valid?
+// If so such an operation would be potentially computationally very expensive.
```
Comment (chfanghr) [inline src/library/plugin/fp/proofCache.ts:48]: > is meant to validate the entire cache given a validator for proofs?

Yes.

> And all proofs that are left after this operation should be regarded as valid?

Yes.

> If so such an operation would be potentially computationally very expensive.

This is true. The reason why it was designed like this is that we only want plugins to invalidate proofs(remove entries from the cache) but not add them.

File: src/library/plugin/fp/proofCache.ts [hunk 10]
```diff
@@ -15,62 +15,112 @@ import {
...
+
+// TODO 2: I think that ProofCache should be redesigned
+// Basically it should be meant as a helper for plugins to allow for their output validation.
+// If we require that each plugin output is uniquely identifiable this can be enough to
+// unlock this possibility.
+/**
+ * A type alias for a callback that takes a cached proof verifier and
```
Comment (chfanghr) [inline src/library/plugin/fp/proofCache.ts:55]: And also note that the function will only be called by the plugin.

**Review by chfanghr — COMMENTED** [review]
Excellent documentation, great thanks!

File: src/library/plugin/fp/proofCache.ts (hunk 9 above)
Comment (chfanghr) [reply src/library/plugin/fp/proofCache.ts:48]: Perhaps allowing plugins to do read-only walk-through and read-only random access is the way to go. Then we can change the invalidate callback to a pure function.

File: src/utils/fp/TaskEither.ts [hunk 11]
```diff
@@ -25,7 +30,12 @@ export function fromFailablePromise<T>(
     );
```
Comment (chfanghr) [inline src/utils/fp/TaskEither.ts:30]: It was overlooked - exceptions won't be handled properly. Could use a `.catch` clause here.

File: src/utils/fp/TaskEither.ts [hunk 12]
```diff
@@ -38,6 +48,11 @@ export function fromFailableVoidPromise(
...
```
Comment (chfanghr) [inline src/utils/fp/TaskEither.ts:48]: ditto

File: src/library/plugin/index.ts (hunk 5 above)
Comment (adamczykm) [reply src/library/plugin/index.ts:72]: Just a change of the location of files. lib/plugin -> lib/server

**Comment by adamczykm** [comment]
Done in #14

**Pull request closed without merging**